lib/spack/spack/package.py

```diff
@@ -1129,6 +1133,34 @@ def do_patch(self):
         else:
             touch(no_patches_file)
 
+    def patches_to_apply(self):
+        """If the patch set does not change between two invocations of spack,
+           then all patches in the set will be applied in the same order"""
+        patchesToApply = itertools.chain.from_iterable(
+            patch_list
+            for spec, patch_list in self.patches.iteritems()
+            if self.spec.satisfies(spec))
+        return sorted(patchesToApply, key=lambda p: p.path_or_url)
+
+    def content_hash(self, remove_pkg_name=False):
+        """Create a hash based on the sources and logic used to build the
+        package. This includes the contents of all applied patches and the
+        contents of applicable functions in the package subclass."""
+        hashContent = list()
+        # TODO? in cases where a digest or source_id isn't available, should
+        # this attempt to download the source and set one? For URL resources
+        # this will involve downloading the entire resource (repositories
+        # typically allow for just retrieving the latest commit ID)
+        source_id = fs.for_package_version(self, self.version).source_id()
+        if not source_id:
+            return
```

Comment:
@scheibelp This causes failures because `package_hash` doesn't expect `None` in its use of this method, so it seems that something will need to be done here.

Reply:
Probably we should make this just hash empty string if we do not have the content.

Reply:
Eh, that seems flaky to me. The hash of a package changes if I delete a patch file on disk?